narwhal/network/src/client.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::BTreeMap, sync::Arc, time::Duration};
+
+use anemo::{PeerId, Request};
+use async_trait::async_trait;
+use crypto::{traits::KeyPair, NetworkKeyPair, NetworkPublicKey};
+use mysten_common::sync::notify_once::NotifyOnce;
+use parking_lot::RwLock;
+use tokio::{select, time::sleep};
+use tracing::debug;
+use types::{
+    error::LocalClientError, PrimaryToWorker, WorkerOthersBatchMessage, WorkerOurBatchMessage,
+    WorkerSynchronizeMessage, WorkerToPrimary,
+};
+
+use crate::traits::{PrimaryToWorkerClient, WorkerToPrimaryClient};
+
+/// NetworkClient provides the interface to send requests to other nodes, and call other components
+/// directly if they live in the same process. It is used by both primary and worker(s).
+///
+/// Currently this only supports local direct calls, and it will be extended to support remote
+/// network calls.
+#[derive(Clone)]
+pub struct NetworkClient {
+    inner: Arc<RwLock<Inner>>,
+    shutdown_notify: Arc<NotifyOnce>,
+}
+
+struct Inner {
+    // The private-public network key pair of this authority.
+    primary_peer_id: PeerId,
+    worker_to_primary_handler: Option<Arc<dyn WorkerToPrimary>>,
+    primary_to_own_worker_handler: BTreeMap<PeerId, Arc<dyn PrimaryToWorker>>,
+    shutdown: bool,
+}
+
+impl NetworkClient {
+    pub fn new(primary_peer_id: PeerId) -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(Inner {
+                primary_peer_id,
+                worker_to_primary_handler: None,
+                primary_to_own_worker_handler: BTreeMap::new(),
+                shutdown: false,
+            })),
+            shutdown_notify: Arc::new(NotifyOnce::new()),
+        }
+    }
+
+    pub fn new_from_keypair(primary_network_keypair: &NetworkKeyPair) -> Self {
+        Self::new(PeerId(primary_network_keypair.public().0.into()))
+    }
+
+    pub fn new_with_empty_id() -> Self {
+        // ED25519_PUBLIC_KEY_LENGTH is 32 bytes.
+        Self::new(empty_peer_id())
+    }
+
+    pub fn set_worker_to_primary_local_handler(&self, handler: Arc<dyn WorkerToPrimary>) {
+        let mut inner = self.inner.write();
+        inner.worker_to_primary_handler = Some(handler);
+    }
+
+    pub fn set_primary_to_worker_local_handler(
+        &self,
+        worker_id: PeerId,
+        handler: Arc<dyn PrimaryToWorker>,
+    ) {
+        let mut inner = self.inner.write();
+        inner
+            .primary_to_own_worker_handler
+            .insert(worker_id, handler);
+    }
+
+    pub fn shutdown(&self) {
+        let mut inner = self.inner.write();
+        if inner.shutdown {
+            return;
+        }
+        inner.worker_to_primary_handler = None;
+        inner.primary_to_own_worker_handler = BTreeMap::new();
+        inner.shutdown = true;
+        let _ = self.shutdown_notify.notify();
+    }
+
+    async fn get_primary_to_worker_handler(
```

Comment:
Modify the getters to include `local` in the name?

Reply:
Leaving `local` out for consistency for now, but we can add it depending on how this evolves.